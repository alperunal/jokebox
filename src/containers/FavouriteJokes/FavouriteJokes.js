import React, { Component } from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Container from '../../components/UI/Container/Container';
import { selectJoke, favouriteJoke } from '../../store/actions/jokeActions';
import JokeDetails from '../../components/JokeDetails/JokeDetails';
import Loader from '../../components/UI/Loader/Loader';
import Modal from '../../components/UI/Modal/Modal';
import Joke from '../../components/Joke/Joke';

export class FavouriteJokes extends Component {
  constructor(props) {
    super(props);
    this.handleModalClose = this.handleModalClose.bind(this);
    this.openJokeDetails = this.openJokeDetails.bind(this);
    this.favouriteJoke = this.favouriteJoke.bind(this);

    this.state = {
      isModalOpen: false,
    };
  }

  handleModalClose() {
    this.setState({ isModalOpen: false });
  }

  openJokeDetails(joke) {
    // eslint-disable-next-line react/destructuring-assignment
    this.props.selectJoke(joke);
    this.setState({ isModalOpen: true });
  }

  favouriteJoke(event, joke) {
    event.stopPropagation();
    // eslint-disable-next-line react/destructuring-assignment
    this.props.favouriteJoke(joke);
  }

  render() {
    const { isModalOpen } = this.state;
    const { favouriteJokes, joke, isLoading } = this.props;

    return (
      <StyledJokes>
        <Container>
          {isLoading && <Loader />}
          {isModalOpen && (
            <Modal closeModal={this.handleModalClose}>
              <JokeDetails joke={joke} />
            </Modal>
          )}
          <Wrapper>
            <Title>Favourite Jokes</Title>
            {favouriteJokes ? (
              <List>
                {favouriteJokes.map(_joke => {
                  return (
                    <ListItemLink key={_joke.id} onClick={() => this.openJokeDetails(_joke)}>
                      <Joke joke={_joke} isFavourite favouriteJoke={this.favouriteJoke} />
                    </ListItemLink>
                  );
                })}
              </List>
            ) : (
              'There is no joke available.'
            )}
          </Wrapper>
        </Container>
      </StyledJokes>
    );
  }
}

const StyledJokes = styled.div`
  display: flex;
  justify-content: center;
`;

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  padding-top: 20px;
  padding-left: 20px;
  padding-right: 20px;
  padding-bottom: 50px;
  background-color: #ffffff;
`;
const Title = styled.div`
  font-size: 20px;
  padding-left: 20px;
  box-shadow: -1px 1px 0 5px #fff, 1px 1px 0 5px #fff, 0 7px 0 0 #e3e3e3;
  margin-bottom: 25px;
`;

const List = styled.div`
  display: grid;
  grid-template-columns: 100%;
  grid-row-gap: 2%;

  // Medium devices (tablets, 768px and up)
  @media (min-width: 768px) {
    grid-template-columns: 45% 45%;
    grid-column-gap: 10%;
    grid-row-gap: 4%;
  }
  // Large devices (desktops, 992px and up)
  @media (min-width: 992px) {
    grid-template-columns: 30% 30% 30%;
    grid-column-gap: 5%;
    grid-row-gap: 4%;
  }
`;

const ListItemLink = styled.span`
  text-decoration: none;
`;

FavouriteJokes.propTypes = {
  selectJoke: PropTypes.func.isRequired,
  favouriteJoke: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  favouriteJokes: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      type: PropTypes.string,
      setup: PropTypes.string,
      punchline: PropTypes.string,
    }),
  ).isRequired,
  joke: PropTypes.shape({
    id: PropTypes.number,
    type: PropTypes.string,
    setup: PropTypes.string,
    punchline: PropTypes.string,
  }).isRequired,
};

const mapStateToProps = state => ({
  favouriteJokes: state.joke.favouriteJokes,
  joke: state.joke.joke,
  isLoading: state.joke.isLoading,
});

export default connect(
  mapStateToProps,
  { selectJoke, favouriteJoke },
)(FavouriteJokes);
