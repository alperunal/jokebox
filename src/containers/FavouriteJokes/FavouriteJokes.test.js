import React from 'react';
import { shallow } from 'enzyme';
import { FavouriteJokes } from './FavouriteJokes';

describe('FavouriteJokes', () => {
  let wrapper;
  let jokes;
  let joke;

  beforeEach(() => {
    jokes = [
      {
        id: 21,
        type: 'general',
        setup: 'A termite walks into a bar and says...',
        punchline: "'Where is the bar tended?'",
      },
      {
        id: 51,
        type: 'general',
        setup: "When a dad drives past a graveyard: Did you know that's a popular cemetery?",
        punchline: 'Yep, people are just dying to get in there',
      },
    ];

    joke = {
      id: 21,
      type: 'general',
      setup: 'A termite walks into a bar and says...',
      punchline: "'Where is the bar tended?'",
    };

    wrapper = shallow(
      <FavouriteJokes
        isLoading={false}
        selectJoke={() => {}}
        favouriteJoke={() => {}}
        joke={joke}
        favouriteJokes={jokes}
        fetchJokes={() => {}}
      />,
    );
  });

  test('renders FavouriteJokes correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('renders Loading correctly', () => {
    wrapper = shallow(
      <FavouriteJokes
        isLoading
        selectJoke={() => {}}
        favouriteJoke={() => {}}
        joke={joke}
        favouriteJokes={jokes}
        fetchJokes={() => {}}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });
});
