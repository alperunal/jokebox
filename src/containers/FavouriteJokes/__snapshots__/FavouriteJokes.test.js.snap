// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavouriteJokes renders FavouriteJokes correctly 1`] = `
<styled.div>
  <Container>
    <styled.div>
      <styled.div>
        Favourite Jokes
      </styled.div>
      <styled.div>
        <styled.span
          key="21"
          onClick={[Function]}
        >
          <Joke
            favouriteJoke={[Function]}
            isFavourite={true}
            joke={
              Object {
                "id": 21,
                "punchline": "'Where is the bar tended?'",
                "setup": "A termite walks into a bar and says...",
                "type": "general",
              }
            }
          />
        </styled.span>
        <styled.span
          key="51"
          onClick={[Function]}
        >
          <Joke
            favouriteJoke={[Function]}
            isFavourite={true}
            joke={
              Object {
                "id": 51,
                "punchline": "Yep, people are just dying to get in there",
                "setup": "When a dad drives past a graveyard: Did you know that's a popular cemetery?",
                "type": "general",
              }
            }
          />
        </styled.span>
      </styled.div>
    </styled.div>
  </Container>
</styled.div>
`;

exports[`FavouriteJokes renders Loading correctly 1`] = `
<styled.div>
  <Container>
    <Loader />
    <styled.div>
      <styled.div>
        Favourite Jokes
      </styled.div>
      <styled.div>
        <styled.span
          key="21"
          onClick={[Function]}
        >
          <Joke
            favouriteJoke={[Function]}
            isFavourite={true}
            joke={
              Object {
                "id": 21,
                "punchline": "'Where is the bar tended?'",
                "setup": "A termite walks into a bar and says...",
                "type": "general",
              }
            }
          />
        </styled.span>
        <styled.span
          key="51"
          onClick={[Function]}
        >
          <Joke
            favouriteJoke={[Function]}
            isFavourite={true}
            joke={
              Object {
                "id": 51,
                "punchline": "Yep, people are just dying to get in there",
                "setup": "When a dad drives past a graveyard: Did you know that's a popular cemetery?",
                "type": "general",
              }
            }
          />
        </styled.span>
      </styled.div>
    </styled.div>
  </Container>
</styled.div>
`;
